### ISSUE TITLE
sftp file uploads/downloads hang or break on RPi, SSHDroid, Remarkable, etc.

### ISSUE THREAD

**Issue by mc-butler** [MidnightCommander/mc#3654]
> [!IMPORTANT]
> This issue was migrated from Trac:
>
> | **Origin** | https://midnight-commander.org/ticket/3654 |
> | :------------- | :------------------------------------------------------------------------------------------------ |
> | **Reporter** | `x905` |

<p>
in 4.8.17 has been broken sftp connection command<br />
server is sshdroid on android device<br />
4.8.17 can connect, browse files and upload small files (up to 20 kb), but 30 kb cant - and hangs<br />
4.8.15 in same situation works fine (except error -31 as always)<br />
</p>
<p>
some add info <a class="ext-link" href="https://www.linux.org.ru/forum/midnight/12592086">https://www.linux.org.ru/forum/midnight/12592086</a><br />
</p>

**Comment by zyv** [comment]
First, a general comment on `libssh2` / `libssh`.

I think that `libssh2` is still actively developed, but it has (or used to have?) issues with modern crypto, and my experience with the interface is not that great, although our usage of it is also quite wrong (see next comment). To my knowledge, for these reasons, Fedora switched to `libssh` for `curl`, and maybe we should too:

https://fedoraproject.org//wiki/Changes/libssh-in-libcurl

My quick appraisal results in the fact that `libssh` API is safer and much easier to use for our purposes:

https://api.libssh.org/master/libssh_tutor_guided_tour.html
https://api.libssh.org/master/libssh_tutor_sftp.html

Specifically, they seem to have a clean synchronous API for SFTP and simple error handling, which is what we need and the lack of which in `libssh2` has been causing most of the issues in the past.

On top of that, I think they have a better `ssh_config` parsing and `known_hosts` support, which would allow us to remove some of our code.

If anyone would like to take on the research, whether it's worth it and solves our (other) longstanding problems (like no proper `known_hosts` interface, etc.), and take up the implementation, especially this time hopefully covered with tests, they're most welcome.

**Comment by zyv** [comment]
Now, to our problems with `libssh2`:

I've looked at it deeper today, and my conclusion is, unfortunately, that our use of it is still a hot mess, irrespective of the issues with the library. Especially, our error handling is sadly still pure madness after many attempts to put lipstick on a pig.

---

Our biggest issue is with the synchronous / asynchronous `libssh2` API usage. According to my understanding, there is a following philosophical difference between the libraries:

* `libssh` wants to provide a clean/simple synchronous interface and has (some) non-blocking functions for performance
* `libssh2` wants to provide a 100% non-blocking interface for performance and can be (mostly) used in a blocking mode

The `libssh2` concept of non-blocking communication works as follows. First, you have to tell the library that you are non-blocking:

```c
libssh2_session_set_blocking(session, 0);
```

After that, any functions can return LIBSSH2_ERROR_EAGAIN and have to be called in a loop like this:

```c
while((rc = libssh2_session_handshake(session, sock)) == LIBSSH2_ERROR_EAGAIN);

if(rc) {
  fprintf(stderr, "Failure establishing SSH session: %d\n", rc);
}
```

But this does not apply to all of them, so every time the documentation has to be checked for the correct usage.

```c
fingerprint = libssh2_hostkey_hash(session, LIBSSH2_HOSTKEY_HASH_SHA1);
``` 

Further, some functions return error codes like above, and some have to be processed with 

```c
do {
  sftp_handle = libssh2_sftp_open(sftp_session, sftppath, …);

  if(!sftp_handle && libssh2_session_last_errno(session) != LIBSSH2_ERROR_EAGAIN) {
    fprintf(stderr, "Unable to open file with SFTP: %ld\n”, libssh2_sftp_last_error(sftp_session));
    goto shutdown;
  }
} while(!sftp_handle);
```

Finally, for some functions, it gets even worse. You must wait on the socket before you call it again, or bad things will happen.

```c
while((nwritten = libssh2_sftp_write(sftp_handle, ptr, nread)) == LIBSSH2_ERROR_EAGAIN) {
  waitsocket(sock, session);
}
```

The documentation says you should actually always do it, but the tutorial doesn’t follow this, for reasons that are not known to me. Maybe they just didn’t want to overcomplicate the code. Maybe in practice, bad things are unlikely to happen for functions other than read/write functions.

> When any of libssh2 functions return LIBSSH2_ERROR_EAGAIN an application should wait for the socket to have data available for reading or writing. Depending on the return value of libssh2_session_block_directions an application should wait for read, write or both.

With this in mind, our current usage of the library is simply plain wrong (and apparently has always been, because the original author put together pieces from tutorials without understanding what he was doing, which has already led to the CVE in the past - #4259).

We start in blocking mode, but then still do `LIBSSH2_ERROR_EAGAIN` loops and socket waits. Of course, because the code is not centralized, we do this inconsistently. Error translation and handling is also applied on an adhoc basis. This causes lockups and wrong behavior that goes unnoticed if the server returns non-fatal errors.

---

Now, unfortunately, even if we clean up the non-blocking mess, there is one more difficulty with `libssh2`. This is the direct reason for complaints in this ticket.

Apparently, for performance reasons, `libssh2_sftp_write` will return a smaller number of bytes written than requested even in synchronous mode:

https://libssh2.org/libssh2_sftp_write.html

> **Write ahead**
> 
> Starting in libssh2 version 1.2.8, the default behavior of libssh2 is to create several smaller outgoing packets for all data you pass to this function and it will return a positive number as soon as the first packet is acknowledged from the server.
> 
> This has the effect that sometimes more data has been sent off but is not acked yet when this function returns, and when this function is subsequently called again to write more data, libssh2 will immediately figure out that the data is already received remotely.
> 
> In most normal situation this should not cause any problems, but it should be noted that if you have once called libssh2_sftp_write() with data and it returns short, you MUST still assume that the rest of the data might have been cached so you need to make sure you do not alter that data and think that the version you have in your next function invoke will be detected or used.
> 
> The reason for this funny behavior is that SFTP can only send 32K data in each packet and it gets all packets acked individually. This means we cannot use a simple serial approach if we want to reach high performance even on high latency connections. And we want that.

This means that the correct way to use the write functions even in blocking mode is as follows:

```c
do {
  nwritten = libssh2_sftp_write(sftp_handle, ptr, nread);

  if(nwritten < 0)
    break;

  ptr += nwritten;
  nread -= (size_t)nwritten;
} while(nread);
```

We don’t do this, and it doesn’t seem to manifest in many cases, but for slow devices and/or networks, it seems to cause these mysterious issues that people are complaining about.

**Comment by zyv** [comment]
So after these long reads @aborodin, this hopefully answers your question above as to what can be done about it... and @fekir now should have some background on how bad the situation really is.

I think it's clear that migrating to `libssh` from `libssh2`, while possibly desirable, is a lot of work, and I can't possibly afford to look into this. It seems to me that to get our code in line with how it should work is a lot less effort with this example and explanations above:

https://libssh2.org/examples/sftp_write.html

We need to do the following steps:

1. Move the `libssh2_session_set_blocking` call to where it should be (after `libssh2_session_init`)
2. Remove the `LIBSSH2_ERROR_EAGAIN` loops and waits on the socket because we are using blocking communication.
3. Turn `sftpfs_waitsocket` into a common error-handling and translation function to translate common `LIBSSH2_FX_*` errors - there is already translation for some errors, but not `ENOSPC` for example.
4. Fix the `libssh2_sftp_write` usage (and also check `libssh2_sftp_read` - I'm not sure whether when reading we just give a size of the buffer and it's okay to underread or not, because retries will happen at a higher level).

Not sure if and when I can ever get to that :( Just clarifying all that took some 3 hours of intensive work!

**Comment by tuffnatty** [comment]
@zyv Yes, I have already read #3654, and have started to improve error handling with the current client, however, I don't have a clear picture how it should look in the end. Is there a vfs you'd recommend me as a reference?

Currently I am trying to convert the extended "SFTP protocol error" codes to errno values so that the outer code could make some use of them. However, the current code displays its own error popup, implements the `ferrno` method by returning the global `errno` variable ignoring the vfs-level `verrno` [removing the method seems to give some immediate improvement], and the global `errno` gets reset during the error popup, and mostly never set on libssh errors.

Some other vfses use me->verrno, some implement a static variable, and it's difficult to understand what design would be accepted.

**Comment by zyv** [comment]
Unfortunately, I'm not familiar with the VFS code myself and can't suggest a well-written VFS which should be used as a reference, if we have one at all. To answer these questions which are absolutely to the point, I'd have to do research myself for which I don't have the capacity at the moment, sadly.

Maybe @mc-worker can share his opinion - he was the one who did the current error handling in SFTP, which was completely broken before his changes. The only thing I can say of the bat is that a static variable is a horrible design, but it was a common approach by C programmers in the old days.
